src/commands.rs

```diff
@@ -0,0 +1,1173 @@
+
+    fn find_add_paths(&mut self, path: &str) -> Result<Vec<UpdateInfo>, Error> {
+        let mut parsed_static_path = StaticPathParser::check(path)?;
+
+        if parsed_static_path.valid != VisitStatus::Valid {
+            return Err("Err: wrong static path".into());
+        }
+        if parsed_static_path.static_path_elements.len() < 2 {
+            return Err("Err: path must end with object key to set".into());
+        }
+
+        let last = parsed_static_path.static_path_elements.pop().unwrap();
+
+        if let StaticPathElement::ObjectKey(key) = last {
+            if let StaticPathElement::Root = parsed_static_path.static_path_elements.last().unwrap()
+            {
+                // Adding to the root
+                Ok(vec![UpdateInfo::AUI(AddUpdateInfo {
+                    path: Vec::new(),
+                    key: key.to_string(),
+                })])
+            } else {
+                // Adding somewhere in existing object, use jsonpath_lib::replace_with
+                let mut selector = Selector::default();
+                if let Err(e) = selector.str_path(
+                    &parsed_static_path
+                        .static_path_elements
+                        .iter()
+                        .map(|e| e.to_string())
+                        .collect::<Vec<String>>()
+                        .join(""),
+                ) {
+                    return Err(e.into());
+                }
+                selector.value(self.val);
+                let mut res = selector.select_with_paths(|_| true)?;
+                Ok(res
+                    .drain(..)
+                    .map(|v| {
+                        UpdateInfo::AUI(AddUpdateInfo {
+                            path: v,
+                            key: key.to_string(),
+                        })
+                    })
+                    .collect())
+            }
+        } else if let StaticPathElement::ArrayIndex(_) = last {
+            // if we reach here with array path we must be out of range
+            // otherwise the path would be valid to be set and we would not
+            // have reached here!!
+            Err("array index out of range".into())
+        } else {
+            Err("path not an object".into())
+        }
+    }
+
+    pub fn find_paths(
+        &mut self,
+        path: &str,
+        option: &SetOptions,
+    ) -> Result<Vec<UpdateInfo>, Error> {
+        if SetOptions::NotExists != *option {
+            let mut selector = Selector::default();
+            let mut res = selector
+                .str_path(path)?
+                .value(self.val)
+                .select_with_paths(|_| true)?;
+            if res.len() > 0 {
+                return Ok(res
+                    .drain(..)
+                    .map(|v| UpdateInfo::SUI(SetUpdateInfo { path: v }))
+                    .collect());
+            }
+        }
+        if SetOptions::AlreadyExists != *option {
+            self.find_add_paths(path)
```

Comment:
If option is `NotExists` we will reach here - may `find_add_paths` find and return an existing value?

Reply:
Yes this only gives you what need to be done, but the WriteHolder implementation will not let you set an existing value in a dictionary so it will be catch after.